### PR TITLE
feat(sdk): allow calling GroupOp.after with multiple ops

### PR DESCRIPTION
**Pull request by Udiknedormin**
Adds variable argument list to `GroupOp`'s `after` method so that it supports calling with multiple arguments while staying backwards-compatible. As per issue #4787.

It seems it could be cherry-picked in the current release branch.

### CODE REVIEW

File: sdk/python/kfp/dsl/_ops_group.py [hunk 1]
```diff
@@ -93,9 +93,11 @@ def __enter__(self):
   def __exit__(self, *args):
     _pipeline.Pipeline.get_default_pipeline().pop_ops_group()
 
-  def after(self, dependency):
-    """Specify explicit dependency on another op."""
+  def after(self, dependency, *ops):
```
Comment (numerology) [inline sdk/python/kfp/dsl/_ops_group.py:96]: For my education, why not merge `dependency` with `ops`?

Comment (Udiknedormin) [reply sdk/python/kfp/dsl/_ops_group.py:96]: As described in [the issue](../issues/4787#issue-746812656), it's for it to remain backwards-compatible with by-keyword call (`group.after(dependency=op)`), considering that it wasn't a positional-only argument. I don't think many users used it, but maybe some did.

It does break no-argument call, which means unpacking a list might throw an error (`group.after(*[])`), but I assume it's less of a problem, considering it's wasn't supported before either.

We can remove it if you think it's not really used by any users.

Comment (Ark-kun) [reply sdk/python/kfp/dsl/_ops_group.py:96]: Thank you for taking the backwards compatibility seriously. This is a great approach.

In this case I think it's pretty safe to make the change. The `OpsGroup` is not considered to be public and the `OpsGroup.after` is likely not used widely. In cases where it's used, it's likely to be used without positional argument.

Comment (Udiknedormin) [reply sdk/python/kfp/dsl/_ops_group.py:96]: @Ark-kun Good to know! If that is the case, then using args directly is probably a better idea, as it's simpler and works safely for any list.

**Comment by Udiknedormin** [comment]
@Ark-kun Following your comment, I simplified the signature and doc string so that it matches the ones in `ContainerOp`.

Speaking of `OpsGroup`: are there any plans to merge `OpsGroup` and `BaseOp` class hierarchies into one class graph? I noticed they are sometimes used in compiler as `Union[OpsGroup, BaseOp]` and also seem to share some functionalities (`name` and `.after(*ops)`). Additionally, all `BaseOp`s and some `OpsGroup`s (`Graph`) do have `inputs` and `outputs`, which may be a starting point for either multi-inheritance or protocols, the latter approach being especially useful considering that generating components from component.yaml may create either DSL objects or specs-and-resolved-graphs, both of which do utilise the same interface (comparing `Graph` and `_ResolvedGraphTask`, `TaskSpec._init_outputs()` and `ContainerOp`).

**Comment by Ark-kun** [comment]
> Speaking of `OpsGroup`: are there any plans to merge `OpsGroup` and `BaseOp` class hierarchies into one class graph? I noticed they are sometimes used in compiler as `Union[OpsGroup, BaseOp]` and also seem to share some functionalities (`name` and `.after(*ops)`). Additionally, all `BaseOp`s and some `OpsGroup`s (`Graph`) do have `inputs` and `outputs`, which may be a starting point for either multi-inheritance or protocols, the latter approach being especially useful considering that generating components from component.yaml may create either DSL objects or specs-and-resolved-graphs, both of which do utilise the same interface (comparing `Graph` and `_ResolvedGraphTask`, `TaskSpec._init_outputs()` and `ContainerOp`).

It would be great to make the DSL and compiler more component-oriented which is relevant to what you talk about. There was also an idea to have `kfp.components.structures.TaskSpec` behind the `kfp.dsl.ContainerOp`. We'll probably need to rewrite the compiler at some point. With the current compiler architecture, it's pretty hard to introduce new DSL features..

**Pull request merged**